### PR TITLE
Added User model and authentication

### PR DESCRIPTION
**Pull request by annaUniversal**
Added User model and authentication

### CODE REVIEW

File: src/models/User.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
+const mongoose = require("mongoose");
+const bcrypt = require("bcryptjs");
+
+const UserSchema = new mongoose.Schema({
+  firstName: {
+    type: String,
+    required: [true, "Please Provide Your First Name"],
+    maxlength: 50,
```
Comment (IrynaKolh) [inline src/models/User.js]: better add err msg [3, "Name must be at least 3 characters"],

File: src/models/User.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    required: [true, "Please Provide Email"],
+    match: [
+      /^([\w-\.]+@([\w-]+\.)+[\w-]{2,4})?$/,
+      "Please Provide a Valid Email",
+    ],
+    unique: true,
+  },
+  password: {
+    type: String,
+    required: [true, "Please provide password"],
```
Comment (IrynaKolh) [inline src/models/User.js:29]: add min lenght for password  minlength: [6, "Password must be at least 6 characters"],

File: src/models/User.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    },
+  },
+  role: {
+    type: String,
+    enum: ["tutor", "parent"]
+  }
+});
+
+UserSchema.pre("save", async function (next) {
+  if (!this.isModified("password")) {
```
Comment (IrynaKolh) [inline src/models/User.js]: as I correct understood this code used for changing password. It is strach goal.

File: src/controllers/auth.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
+const User = require("../models/User");
+const { StatusCodes } = require("http-status-codes");
+const { BadRequestError, UnauthenticatedError } = require("../errors");
+
+//const jwt = require("jsonwebtoken");
+
+const register = async (req, res) => {
+  // can do validation vis mangoos in the controller dirrectly
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: delete not needed comments

File: src/controllers/auth.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+  // }
+
+  const user = await User.create({ ...req.body });
+  //const token = jwt.sign({ userId: user._id, name: user.name }, "jwtSecret", {
+  //   expiresIn: "30d",
+  // }); //user.createJWT()
+
+  console.log("test");
+  console.log("name ", user.name);
+  const token = user.createJWT();
```
Comment (IrynaKolh) [inline src/controllers/auth.js:12]: add createJWT function to User model file

File: src/controllers/auth.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  // const token = user.createJWT();
+  console.log("token ", token);
+  res.status(StatusCodes.CREATED).json({ user: { name: user.name }, token });
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: return user role too

File: src/middleware/authentification.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+  }
+
+  const token = authHeader.split(" ")[1];
+
+  try {
+    const payload = jwt.verify(token, process.env.JWY_SECRET);
+    // attach user to the job rout
+    req.user = { userId: payload.userId, name: payload.name };
+
+    // can be used the following
```
Comment (IrynaKolh) [inline src/middleware/authentification.js]: delete not needed comments

File: src/models/User.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  }
+  try {
+    const salt = await bcrypt.genSalt(10);
+    this.password = await bcrypt.hash(this.password, salt);
+    next();
+  } catch (error) {
+    next(error);
+  }
+});
+
```
Comment (IrynaKolh) [inline src/models/User.js:63]: add createJWT function to this file

File: .gitignore [hunk 9]
```diff
@@ -15,5 +15,19 @@ node_modules/
...
\ No newline at end of file
+.DS_Store
+.env.local
+.env.development.local
+.env.test.local
+.env.production.local
+
+npm-debug.log*
+yarn-debug.log*
+yarn-error.log*
```
Comment (IrynaKolh) [inline .gitignore:34]: add package-lock.json to .gitignore

File: src/middleware/storeLocals.js [hunk 10]
```diff
@@ -0,0 +1,12 @@
+const storeLocals = (req, res, next) => {
```
Comment (IrynaKolh) [inline src/middleware/storeLocals.js]: where do you use this file?

Comment (annaUniversal) [reply src/middleware/storeLocals.js]: No. Used it for ejs. Then remembered that we don't use it and didn't delete.
Will clean up

File: src/models/User.js (hunk 2 above)
Comment (annaUniversal) [reply src/models/User.js:29]: added

File: src/models/User.js (hunk 3 above)
Comment (annaUniversal) [reply src/models/User.js]: It is. I'll keep it. If we wouldn't have time for it, I'll remove it

File: src/controllers/auth.js (hunk 5 above)
Comment (annaUniversal) [reply src/controllers/auth.js:12]: added

File: src/controllers/auth.js (hunk 6 above)
Comment (annaUniversal) [reply src/controllers/auth.js]: added

File: src/models/User.js (hunk 8 above)
Comment (annaUniversal) [reply src/models/User.js:63]: added for the earlier comment

File: src/middleware/authentification.js [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+  // check header
+  const authHeader = req.headers.authorization;
+  if (!authHeader || !authHeader.startsWith("Bearer ")) {
+    throw new UnauthenticatedError("Authorization invalid");
+  }
+
+  const token = authHeader.split(" ")[1];
+
+  try {
+    const payload = jwt.verify(token, process.env.JWY_SECRET);
```
Comment (oxangyal) [inline src/middleware/authentification.js]: maybe JWT not JWY?

File: src/models/User.js
Comment (oxangyal) [inline src/models/User.js:1]: you forgot to import token
const jwt = require('jsonwebtoken');

File: src/middleware/authentification.js (hunk 11 above)
Comment (annaUniversal) [reply src/middleware/authentification.js]: Yes. Typo

File: .gitignore (hunk 9 above)
Comment (annaUniversal) [reply .gitignore:34]: added

File: src/controllers/auth.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+const { StatusCodes } = require("http-status-codes");
+const { BadRequestError, UnauthenticatedError } = require("../errors");
+
+const jwt = require("jsonwebtoken");
+
+const register = async (req, res) => {
+  const user = await User.create({ ...req.body });
+
+  console.log("test");
+  console.log("name ", user.name);
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: we don't have name, we have first and last name.

 console.log("firstName ", user.firstName, "lastName ", user.lastName, "email ", user.email, "password ", user.password, "role ", user.role);  - full info about user can print like this

File: src/controllers/auth.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  const token = user.createJWT();
+  console.log("token ", token);
+  res.status(StatusCodes.CREATED).json({ user: { name: user.name, role: user.role }, token });
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: res.status(StatusCodes.CREATED).json({ user: { firstName: user.firstName, lastName: user.lastName, role: user.role }, token });

File: src/controllers/auth.js [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+  const isPasswordCorrect =await user.comparePassword(password)
+  if (!isPasswordCorrect) {
+    throw new UnauthenticatedError("The access is not authorised");
+  } 
+
+  //comparing password
+  const token = user.createJWT();
+  console.log("userId ", user._id)
+  res.status(StatusCodes.OK).json({ user: { name: user.name }, token });
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: res.status(StatusCodes.OK).json({ user: { firstName: user.firstName, lastName: user.lastName, role: user.role }, token });

File: src/controllers/auth.js [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+};
+
+
+const empt =async (req, res) => {
```
Comment (IrynaKolh) [inline src/controllers/auth.js]: delete this

File: src/models/User.js [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  firstName: {
+    type: String,
+    required: [true, "Please Provide Your First Name"],
+    maxlength: [50,"First name should not be more that 50 characters"],
+    minlength: [3, "First name must be at least 3 characters"],
+  },
+  lastName: {
+    type: String,
+    required: [true, "Please Provide Your Last Name"],
+    maxlength: 50,
```
Comment (IrynaKolh) [inline src/models/User.js]: maxlength: [50,"Last name should not be more that 50 characters"],

File: src/routes/auth.js [hunk 17]
```diff
@@ -0,0 +1,10 @@
+const express = require("express");
+const router = express.Router();
+
+const { login, register, empt } = require("../controllers/auth");
+
+router.post("/register", register);
+router.post("/login", login);
+router.post("/empt", empt);
```
Comment (IrynaKolh) [inline src/routes/auth.js]: delete route

**Review by IrynaKolh — COMMENTED** [review]
fix small bags and we can merge this code to main)))

File: src/models/User.js [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+    next(error);
+  }
+});
+
+UserSchema.methods.comparePassword = async function (candidatePassword) {
+  return await bcrypt.compare(candidatePassword, this.password);
+};
+
+UserSchema.methods.createJWT = function () {
+    const token = jwt.sign({ userId: this._id, name: this.firstName }, process.env.JWT_SECRET, {
```
Comment (IrynaKolh) [inline src/models/User.js]: const token = jwt.sign({ userId: this._id, firstName: this.firstName, lastName: this.lastName }, process.env.JWT_SECRET, {
      expiresIn: "30d",
    });

File: src/middleware/authentication.js [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+  if (!authHeader || !authHeader.startsWith("Bearer ")) {
+    throw new UnauthenticatedError("Authorization invalid");
+  }
+
+  const token = authHeader.split(" ")[1];
+
+  try {
+    const payload = jwt.verify(token, process.env.JWT_SECRET);
+    // attach user to the job rout
+    req.user = { userId: payload.userId, name: payload.name };
```
Comment (IrynaKolh) [inline src/middleware/authentication.js]: req.user = { userId: payload.userId, firstName: payload.firstName, lastName: payload.lastName};

File: src/models/User.js (hunk 18 above)
Comment (annaUniversal) [reply src/models/User.js]: Done

File: src/routes/auth.js (hunk 17 above)
Comment (annaUniversal) [reply src/routes/auth.js]: Done

File: src/models/User.js (hunk 16 above)
Comment (annaUniversal) [reply src/models/User.js]: Done

File: src/controllers/auth.js (hunk 15 above)
Comment (annaUniversal) [reply src/controllers/auth.js]: Done

File: src/controllers/auth.js (hunk 14 above)
Comment (annaUniversal) [reply src/controllers/auth.js]: Done

File: src/controllers/auth.js (hunk 13 above)
Comment (annaUniversal) [reply src/controllers/auth.js]: Done

File: src/controllers/auth.js (hunk 12 above)
Comment (annaUniversal) [reply src/controllers/auth.js]: Done

File: src/middleware/authentication.js (hunk 19 above)
Comment (annaUniversal) [reply src/middleware/authentication.js]: Done

**Review by IrynaKolh — APPROVED** [review]
LGFM

**Review by oxangyal — APPROVED** [review]
Looks good!

**Pull request merged**